File: Android/src/app/src/main/java/com/google/ai/edge/gallery/customtasks/voiceassistant/EmotionText.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ai.edge.gallery.customtasks.voiceassistant

/**
 * A one-shot cue to play an emotion effect (floating emoji) on the chat screen, emitted when an
 * assistant reply contains emoji. The [id] makes each cue unique so identical emoji re-trigger.
 */
data class EmotionCue(val emojis: List<String>, val id: Long)

// Broad set of emoji / pictographic / symbol ranges (plus variation selectors, ZWJ, skin tones,
// keycap) — stripped from text before TTS so the engine doesn't read them aloud awkwardly.
private val SPEECH_SYMBOL_RANGES =
  "\\x{1F000}-\\x{1FAFF}" +
    "\\x{2600}-\\x{27BF}" +
    "\\x{2B00}-\\x{2BFF}" +
    "\\x{1F1E6}-\\x{1F1FF}" +
    "\\x{2300}-\\x{23FF}" +
    "\\x{2500}-\\x{25FF}" +
    "\\x{2190}-\\x{21FF}" +
    "\\x{1F3FB}-\\x{1F3FF}" +
    "\\x{FE00}-\\x{FE0F}" +
    "\\x{200D}" +
    "\\x{20E3}"

private val speechSymbolRegex = Regex("[$SPEECH_SYMBOL_RANGES]")
// Markdown / read-aloud-unfriendly punctuation to drop (keeps normal sentence punctuation).
private val markdownRegex = Regex("[*_#`~|>\\\\^•]")
private val whitespaceRegex = Regex("\\s+")

// A complete fenced code block: ```lang\n ... ``` (optional language, multiline body). Removed from
// spoken text entirely so the engine never reads code aloud.
private val fencedCodeRegex = Regex("```[\\s\\S]*?```")
// Inline code: `code` (single backticks, no newline inside).
private val inlineCodeRegex = Regex("`[^`\\n]*`")

/**
 * Removes code from [text] so it is never read aloud: complete fenced code blocks and inline code
 * are dropped. Used by both the after-complete and streaming TTS paths. The on-screen text is left
 * untouched — only the spoken version loses the code.
 */
fun stripCodeForSpeech(text: String): String =
  text.replace(fencedCodeRegex, " ").replace(inlineCodeRegex, " ")

/**
 * Returns the portion of a still-streaming reply [full] that is safe to speak: complete fenced code
 * blocks are removed, and an *open* (not-yet-closed) fence — plus everything after it — is held back
 * so a half-streamed code block is never spoken. Once the fence closes it is dropped by
 * [stripCodeForSpeech]. The returned prefix is stable across calls (closed/removed blocks don't shift
 * the already-spoken prefix), so callers can track spoken position against it.
 */
fun speakableStreamingView(full: String): String {
  // Count fences; an odd count means the last one is still open.
  val withoutClosed = full.replace(fencedCodeRegex, " ")
  val openFence = withoutClosed.indexOf("```")
  val held = if (openFence >= 0) withoutClosed.substring(0, openFence) else withoutClosed
  return held.replace(inlineCodeRegex, " ")
}

/**
 * Returns [text] cleaned for text-to-speech: code, emoji, pictographs and markdown symbols removed
 * and whitespace collapsed, so the spoken reply sounds natural. May return an empty string (e.g. a
 * reply that was only code or an emoji) — callers should skip speaking in that case.
 */
fun sanitizeForSpeech(text: String): String =
  stripCodeForSpeech(text)
    .replace(speechSymbolRegex, " ")
    .replace(markdownRegex, " ")
    .replace(whitespaceRegex, " ")
    .trim()

// Conservative emoji ranges used to drive the visual effect (avoids arrows / geometric symbols).
private val EMOJI_ANIM_RANGES =
  "\\x{1F300}-\\x{1FAFF}" + "\\x{2600}-\\x{27BF}" + "\\x{2B00}-\\x{2BFF}" + "\\x{1F1E6}-\\x{1F1FF}"

private val emojiClusterRegex =
  Regex("[$EMOJI_ANIM_RANGES](?:[\\x{1F3FB}-\\x{1F3FF}\\x{FE0F}\\x{200D}$EMOJI_ANIM_RANGES])*")

/** Extracts emoji (as grapheme-ish clusters) from [text], for the emotion effect. */
fun extractEmojis(text: String): List<String> =
  emojiClusterRegex.findAll(text).map { it.value }.filter { it.isNotBlank() }.toList()

File: iOS/Gallery/CustomTasks/VoiceAssistant/EmotionText.swift
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Port of customtasks/voiceassistant/EmotionText.kt

import Foundation

/// A one-shot cue to play an emotion effect (floating emoji) on the chat screen,
/// emitted when an assistant reply contains emoji. The `id` makes each cue unique
/// so identical emoji re-trigger the animation.
struct EmotionCue {
    let emojis: [String]
    let id: Int64
}

// MARK: - TTS text sanitization

/// Removes fenced and inline code blocks from `text` so they are never read aloud.
func stripCodeForSpeech(_ text: String) -> String {
    let fenced = try? NSRegularExpression(pattern: "```[\\s\\S]*?```", options: [])
    let inline = try? NSRegularExpression(pattern: "`[^`\\n]*`", options: [])
    var result = text
    let range = NSRange(result.startIndex..., in: result)
    result = fenced?.stringByReplacingMatches(in: result, range: range, withTemplate: " ") ?? result
    let range2 = NSRange(result.startIndex..., in: result)
    result = inline?.stringByReplacingMatches(in: result, range: range2, withTemplate: " ") ?? result
    return result
}

/// Returns the part of a still-streaming reply that is safe to speak:
/// complete fenced code blocks are removed; an open (not-yet-closed) fence —
/// plus everything after it — is held back. Mirrors `speakableStreamingView`.
func speakableStreamingView(_ full: String) -> String {
    let fenced = try? NSRegularExpression(pattern: "```[\\s\\S]*?```", options: [])
    let range = NSRange(full.startIndex..., in: full)
    var withoutClosed = fenced?.stringByReplacingMatches(in: full, range: range, withTemplate: " ") ?? full
    if let openRange = withoutClosed.range(of: "```") {
        withoutClosed = String(withoutClosed[withoutClosed.startIndex..<openRange.lowerBound])
    }
    let inlineRange = NSRange(withoutClosed.startIndex..., in: withoutClosed)
    let inline = try? NSRegularExpression(pattern: "`[^`\\n]*`", options: [])
    return inline?.stringByReplacingMatches(in: withoutClosed, range: inlineRange, withTemplate: " ") ?? withoutClosed
}

/// Returns `text` cleaned for TTS: code, emoji, pictographs and markdown removed
/// and whitespace collapsed. May return an empty string.
func sanitizeForSpeech(_ text: String) -> String {
    var result = stripCodeForSpeech(text)
    // Strip emoji/pictographic Unicode ranges. Swift's Character.isEmoji would
    // miss some ranges; we remove by Unicode scalar value instead.
    result = result.unicodeScalars.filter { scalar in
        let v = scalar.value
        // Emoji + symbols ranges (mirrors Android SPEECH_SYMBOL_RANGES).
        let isSymbol =
            (v >= 0x1F000 && v <= 0x1FAFF) ||
            (v >= 0x2600  && v <= 0x27BF)  ||
            (v >= 0x2B00  && v <= 0x2BFF)  ||
            (v >= 0x1F1E6 && v <= 0x1F1FF) ||
            (v >= 0x2300  && v <= 0x23FF)  ||
            (v >= 0x2500  && v <= 0x25FF)  ||
            (v >= 0x2190  && v <= 0x21FF)  ||
            (v >= 0x1F3FB && v <= 0x1F3FF) ||
            (v >= 0xFE00  && v <= 0xFE0F)  ||
            v == 0x200D || v == 0x20E3
        return !isSymbol
    }.reduce("") { $0 + String($1) }
    // Strip markdown punctuation that sounds odd when read aloud.
    let mdRegex = try? NSRegularExpression(pattern: "[*_#`~|>\\\\^•]", options: [])
    let r = NSRange(result.startIndex..., in: result)
    result = mdRegex?.stringByReplacingMatches(in: result, range: r, withTemplate: " ") ?? result
    // Collapse whitespace.
    let wsRegex = try? NSRegularExpression(pattern: "\\s+", options: [])
    let r2 = NSRange(result.startIndex..., in: result)
    result = wsRegex?.stringByReplacingMatches(in: result, range: r2, withTemplate: " ") ?? result
    return result.trimmingCharacters(in: .whitespaces)
}

// MARK: - Emoji extraction for the visual emotion effect

/// Extracts emoji clusters from `text` for the floating emotion animation.
func extractEmojis(_ text: String) -> [String] {
    // Conservative emoji ranges (avoids arrows / geometric symbols).
    let pattern = "[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{2B00}-\u{2BFF}\u{1F1E6}-\u{1F1FF}]" +
                  "(?:[\u{1F3FB}-\u{1F3FF}\u{FE0F}\u{200D}" +
                  "\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{2B00}-\u{2BFF}\u{1F1E6}-\u{1F1FF}])*"
    guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else { return [] }
    let range = NSRange(text.startIndex..., in: text)
    return regex.matches(in: text, range: range).compactMap { match in
        Range(match.range, in: text).map { String(text[$0]) }
    }.filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
}
